Handle nested transactions

piRGoif opened the pull request:
Ticket N°679 was aiming at adding more reliability on object insertion by adding a transaction around INSERT statements for an object (there is one table per class in the object hierarchy).

But we already have transactions in iTop, especially in datasynchro : when simulating, a transaction is started and rollbacked at the end.

MySQL doesn't support nested transactions, and indeed every START TRANSACTION will automatically COMMIT the previous started transaction ! (see [official doc](https://dev.mysql.com/doc/refman/5.6/en/commit.html))

I started implementing a solution using savepoints, that would rollback only the nested transaction if rollback is called.
But after a chat with Denis and Bruno, we decided to have a simpler solution : just one transaction, and :

* if calling COMMIT then sends to db if we're at the root level, otherwise do nothing
* if calling ROLLBACK : sends to db if we're at the root level, else throw a new MySQLChildTransactionRollbackException

piRGoif committed:
N°679 CMDBSource now handles nested transactions

piRGoif committed:
N°679 Transaction handling : allow to force rollback, and handle commit with empty transaction stack

piRGoif committed:
N°679 Handles nested transaction by having only one global transaction

bruno-ds commented on core/cmdbsource.class.inc.php:
```diff
@@ -99,6 +99,15 @@ public function __construct($sIssue, $aContext)
+ * @see \CMDBSource::Rollback()
+ * @since 2.7.0 N°679
+ */
+class MySQLChildTransactionRollbackException extends MySQLException
```
You know I'm paranoid:
Shouldn't we extend \Exception in order to prevent some possible catch of \MySQLException?
(At least a check on the codebase would be reassuring)

piRGoif replied:
Yeah, I asked myself the same question... But I didn't find in corresponding code some catch of MySQLException so I go on implementing a child MySQLException class...

piRGoif replied:
See comment below, after a meeting with Denis & Eric

bruno-ds commented on core/cmdbsource.class.inc.php:
```diff
@@ -130,6 +139,12 @@ class CMDBSource
+	 * @var string[] unique identifier for each nested transaction level (array[0] => 1st level, array[1] => sub transaction, etc)
+	 * @since 2.7.0 N°679
+	 */
+	protected static $m_aTransacSavepoints = array();
```
To be removed?

piRGoif replied:
I'm closing this as the same subject is listed below :)

bruno-ds commented on core/cmdbsource.class.inc.php:
```diff
@@ -570,25 +585,64 @@ public static function Quote($value, $bAlways = false, $cQuoteStyle = "'")
+	{
+		if (preg_match('/^START TRANSACTION;?$/i', $sSQLQuery))
+		{
+			self::StartTransaction();
```
Maybe we could @trigger_error E_USER_DEPRECATED (I'm not sure about the constant real name)

piRGoif replied:
See comment below, after a meeting with Denis & Eric

bruno-ds commented on core/cmdbsource.class.inc.php:
```diff
@@ -570,25 +585,64 @@ public static function Quote($value, $bAlways = false, $cQuoteStyle = "'")
+	 * @throws \MySQLException
+	 * @throws \MySQLHasGoneAwayException
+	 */
+	public static function DBQuery($sSql)
```
Maybe it could be protected/private?

piRGoif replied:
It was private in the first implementation but I changed it to public after our discussion with Denis...

bruno-ds commented on core/cmdbsource.class.inc.php:
```diff
@@ -599,10 +653,140 @@ public static function Query($sSQLQuery)
+	 *
+	 * @since 2.7.0 N°679
+	 */
+	private static function AddTransactionLevel($bHasExistingTransactions)
```
If the savepoints are removed, then this method should be simplified

bruno-ds replied:
In fact using a counter rather than an array seems simpler

piRGoif replied:
I didn't removed all of the savepoint implementation, I thought it might be useful but you're right, should do one thing at a time. Will change this to a simple counter as we discussed.

piRGoif replied:
It's now a counter with 71ed546f

bruno-ds reviewed (commented):
Maybe the three *TransactionLevel()
Should be renamed if we move from the array to a more simple counter.

bruno-ds commented:
The "force" option for the rollback make me wonder if it is just smart or yagni

piRGoif commented:
> The "force" option for the rollback make me wonder if it is just smart or yagni

I added the force parameter as before the DBQuery method was private. Now it is public, so you can still query directly the DB and send a Rollback. But I prefer to have this clear way of doing a rollback anyway, I think it is simpler for the developer ?

piRGoif commented on core/cmdbsource.class.inc.php:
```diff
@@ -599,10 +653,140 @@ public static function Query($sSQLQuery)
+	 * @throws \MySQLHasGoneAwayException
+	 * @since 2.7.0 N°679
+	 */
+	public static function Rollback($bForce = false)
```
Note for monday : should add an Exception parameter, as Rollback is called in a catch block. Passing the original exception would allow to keep it in the throw stack.

piRGoif replied:
Done with 18387e00

piRGoif committed:
N°679 Transaction level : simplify !
Now uses a simple counter instead of an array (last one was useful for savepoints, but we don't need it anymore)

piRGoif committed:
N°679 Use new methods on existing transactions in iTop
* add new Exception parameter in Rollback()
* change calls from Query() to corresponding dedicated transactions methods

piRGoif commented:
> Maybe the three *TransactionLevel()
> Should be renamed if we move from the array to a more simple counter.

I like to have some indirection so that each consumer doesn't have to worry about the implementation. Indeed, the "transaction level" mechanism should be an object with an interface... But this might be overkill :o)

piRGoif commented:
Here are the conclusion after a chat with @dflaven and @eespie :

* decided to change visibility of \CMDBSource::DBQuery to private (was here to provide a workaround if developer would be blocked, but we are confident in our code)
* all of the new transaction dedicated methods will become private too (don't need to change the consumer, don't need to call them directly)
* we fear that adding transactions would bring some problems, maybe on very specific infrastructure (cluster for example) : we decided to add a new config flag that would allow to disable the new transactions in \DBObject::DBInsertNoReload in case such problem would become blocker

piRGoif committed:
N°679 Finalize nested transactions handling
* forgot to rename \CMDBSource::$m_iTransactionLevel (woops)
* \CMDBSource::Rollback does not send an exception anymore (as rollback is called mainly in catch blocks with existing code after the call...)
* change transaction methods visibility : hides \CMDBSource::DBQuery and all of the transaction handling dedicated methods but \CMDBSource::IsInsideTransaction
* do not change current consumers (reverts commit 18387e001edaff22c3a40ca9e33b7b1b730c0bfb)

piRGoif committed:
N°679 new Config flag to disable transactions in core

piRGoif commented:
Modifications have been pushed a few minutes ago. Time to merge !

piRGoif committed:
N°679 Rollback is now private, so force parameter is not needed anymore
